demos/guide/quic-hq-interop.c

```diff
@@ -0,0 +1,1109 @@
+     * needed.
+     */
+    BIO_set_fd(bio, sock, BIO_CLOSE);
+
+    return bio;
+}
+
+/**
+ * @brief Waits for activity on the SSL socket, either for reading or writing.
+ *
+ * This function monitors the underlying file descriptor of the given SSL
+ * connection to determine when it is ready for reading or writing, or both.
+ * It uses the select function to wait until the socket is either readable
+ * or writable, depending on what the SSL connection requires.
+ *
+ * @param ssl A pointer to the SSL object representing the connection.
+ *
+ * @note This function blocks until there is activity on the socket or
+ * until the timeout specified by OpenSSL is reached. In a real application,
+ * you might want to perform other tasks while waiting, such as updating a
+ * GUI or handling other connections.
+ *
+ * @note This function uses select for simplicity and portability. Depending
+ * on your application's requirements, you might consider using other
+ * mechanisms like poll or epoll for handling multiple file descriptors.
+ */
+static void wait_for_activity(SSL *ssl)
+{
+    fd_set wfds, rfds;
+    int width, sock, isinfinite;
+    struct timeval tv;
+    struct timeval *tvp = NULL;
+
+    /* Get hold of the underlying file descriptor for the socket */
+    sock = SSL_get_fd(ssl);
+
+    FD_ZERO(&wfds);
+    FD_ZERO(&rfds);
+
+    /*
+     * Find out if we would like to write to the socket, or read from it (or
+     * both)
+     */
+    if (SSL_net_write_desired(ssl))
+        FD_SET(sock, &wfds);
+    if (SSL_net_read_desired(ssl))
+        FD_SET(sock, &rfds);
+    width = sock + 1;
+
+    /*
+     * Find out when OpenSSL would next like to be called, regardless of
+     * whether the state of the underlying socket has changed or not.
+     */
+    if (SSL_get_event_timeout(ssl, &tv, &isinfinite) && !isinfinite)
+        tvp = &tv;
+
+    /*
+     * Wait until the socket is writeable or readable. We use select here
+     * for the sake of simplicity and portability, but you could equally use
+     * poll/epoll or similar functions
+     *
+     * NOTE: For the purposes of this demonstration code this effectively
+     * makes this demo block until it has something more useful to do. In a
+     * real application you probably want to go and do other work here (e.g.
+     * update a GUI, or service other connections).
+     *
+     * Let's say for example that you want to update the progress counter on
+     * a GUI every 100ms. One way to do that would be to use the timeout in
+     * the last parameter to "select" below. If the tvp value is greater
+     * than 100ms then use 100ms instead. Then, when select returns, you
+     * check if it did so because of activity on the file descriptors or
+     * because of the timeout. If the 100ms GUI timeout has expired but the
+     * tvp timeout has not then go and update the GUI and then restart the
+     * "select" (with updated timeouts).
+     */
```

Comment:
Probably not worth going there for now, but `select()` is a problematic interface, and one would typically use `poll(2)` on Unix, (or epoll, ...), and some appropriate Windows API on Windows. And of course there'd be a global event loop, that services both the socket and the GUI events, ...

Reply:
yeah, this was taken verbatim from the quic client.  I'm all for using (e)poll here, but I'm a bit hesitant to do that as , as you note, windows requires a separate implementation for that, since it lacks epoll.  Note taken though to improve in a subsequent pr.

Reply:
comments updated to reflect reasoning behind the use of select here and why a "real" application may choose to do otherwise